bin/studio.py

```diff
@@ -0,0 +1,276 @@
+    studio_cli.build_commands(subparsers)
+
+    if not unprocessed_args:
+        parser.print_help()
+        sys.exit(1)
+
+    args = parser.parse_args()
+
+    command = unprocessed_args[0]
+
+    response = studio_cli.execute(command, args)
+
+    print(response)
+
+
+class StudioCli:
+    def __init__(self, config):
+        self.public_api_client = PublicAPIClient(config)
+        self.schema = self.get_schema()
+        self.commands = {}
+
+    # json at
+    # https://tw.instructuremedia.com/api/public/apidocs
+    def get_schema(self):
+        response = self.public_api_client.request("get", "apidocs", version_prefix="")
+
+        if response.status_code != 200:
+            raise Exception(
+                f"Could not get details for {response.url}: {response.text}"
+            )
+
+        return response.json()
+
+    type_to_type = {
+        "integer": int,
+        "string": str,
+        # not sure about these
+        "object": str,
+        "array": str,
+        "file": str,
+    }
+
+    def build_commands(self, subparsers):
+        for path, methods in self.schema["paths"].items():
+            for method, data in methods.items():
+                command = Command(path, method, data, self.public_api_client)
+
+                missing_params = [
+                    parameter
+                    for parameter in command.path_params
+                    if parameter not in command.parameter_names()
+                ]
+
+                if missing_params:
+                    sys.stderr.write(
+                        f"Skipping {command.name} for {path}, params not defined: {missing_params}\n"
+                    )
+                    continue
+
+                parameters = data.get("parameters", {})
+
+                if any([param['in'] in ['body', 'formData'] for param in parameters]):
+                    # commands in body parameters are not supported atm.
+                    # same for file/formData parameters
+                    continue
+
+                command_parser = subparsers.add_parser(
+                    command.name, help=data["summary"]
+                )
+
+                self.commands[command.name] = command
+
+                for param in parameters:
+                    command_parser.add_argument(
+                        f"--{param['name']}",
+                        type=StudioCli.type_to_type[param.get("type", "string")],
+                        required=param.get("required", False),
+                        choices=param.get("enum", None),
+                        help=param["description"],
+                    )
+
+    def execute(self, command_name, args):
+        command = self.commands[command_name]
+        return command.execute(args)
+
+
+class Command:
+    method_to_command = {
+        "delete": "delete",
+        "get": "show",
+        "post": "add",
+        "put": "update",
+    }
+
+    def __init__(self, path, method, data, public_api_client) -> None:
```

Comment:
Do we need ` -> None`? This is the only type hint we use.

Reply:
Bump.

Reply:
Removed it.